### PR TITLE
Move kNN search and dense vectors to core

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR proposes moving kNN search and dense vector support out of an xpack
plugin and into server.

In #87625 we plan to integrate ANN search into the main `_search` endpoint as a
new top-level component called `knn`. So kNN will be a dedicated part of the
search request, and we'll have kNN logic within the search phases. The classes
and logic will live in server, matching the other search components like
suggesters, field collapsing, etc.

Why can't we leave dense vector in its own plugin, and just add kNN search
logic in server?
* This makes testing more difficult/ confusing, since kNN search can only be
exercised effectively using dense vector fields. All tests for kNN search would
need to live in the dense vector plugin, although kNN is part of core search.
* It feels cleanest for all logic related to kNN search to live in the same
place.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java [hunk 1]
```diff
@@ -257,8 +258,8 @@ public PostingsFormat getPostingsFormat(String field) {
      */
     public KnnVectorsFormat getKnnVectorsFormatForField(String field) {
         Mapper fieldMapper = fieldMappers.get(field);
-        if (fieldMapper instanceof PerFieldKnnVectorsFormatFieldMapper) {
-            return ((PerFieldKnnVectorsFormatFieldMapper) fieldMapper).getKnnVectorsFormatForField();
+        if (fieldMapper instanceof DenseVectorFieldMapper) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:261]: drive-by comment, that has little to do with this change specifically, as this logic was already here. Do we need an additional public method in MappingLookup for this? Would it make sense to streamline this into the Mapper base class to be overridden by DenseVectorFieldMapper?

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:261]: Yes, I'm not sure why I thought we needed this! I can remove that `MappingLookup` method now. I am not sure if we want a new top-level method `Mapper#getKnnVectorsFormatForField`, as opposed just doing an `instanceof` check for `DenseVectorFieldMapper`? My inspiration for this change was `CompletionFieldMapper`, which has a special package-private method to define `PostingsFormat`.

**Review by mayya-sharipova — APPROVED** [review]
@jtibshirani Thanks, this arrangement LGTM.

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java]: thanks for the background Alan, makes sense to me as well.

File: modules/lang-painless/src/main/generated/whitelist-json/painless-score.json [hunk 2]
```diff
@@ -1 +1 @@
...
...tUtils$DecayNumericLinear","name":"decayNumericLinear","return":"double","read_only":4,"parameters":["double","double","double","double","double"]},{"declaring":"org.elasticsearch.script.VectorScoreScriptUtils$DotProduct","name":"dotProduct","return":"double","read_only":3,"parameters":["org.elasticsearch.script.ScoreScript","java.util.List","java.lang.String"]},{"declaring":"org.elasticsearch.script.VectorScoreScriptUtils$L1Norm","name":"l1norm","return":"double","read_only":3,"parameters":["org.elasticsearch.script.ScoreScript","java.util.List","java.lang.String"]},{"declaring":"org.elasticsearch.script.VectorScoreScriptUtils$L2Norm","name":"l2norm","return":"double","read_only":3,"parameters":["org.elasticsearch.script.ScoreScript","java.util.List","java.lang.String"]},{"declaring":"org.elasticsearch.script.ScoreScriptUtils$RandomScoreDoc","name":"randomScore","return":"double","read_only":2,"parameters":["org.elasticsearch.script.ScoreScript","int"]},{"declaring":"org.elasticsearch.script.ScoreScriptUtils$RandomScoreField","name":"randomScore","return":"double","read_only":3,"parameters":["org.elasticsearch.script.ScoreScript","int","java.lang.String"]}],"instance_bindings":[]}
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply modules/lang-painless/src/main/generated/whitelist-json/painless-score.json:1]: Some of the full class names changed they were renamed, like `org.elasticsearch.script.VectorScoreScriptUtils$CosineSimilarity`.
